### PR TITLE
Add interface for MD5 plugins (DB-126)

### PR DESCRIPTION
**Pull request by shaan1337**

### CODE REVIEW

File: src/EventStore.Plugins/Cryptography/ICryptoPlugin.cs [hunk 1]
```diff
@@ -0,0 +1,13 @@
+﻿namespace EventStore.Plugins.Cryptography;
+
+public interface ICryptoPlugin {
```
Comment (timothycoleman) [inline src/EventStore.Plugins/Cryptography/ICryptoPlugin.cs]: on names, perhaps `IHashAlgorithmPlugin`? open to suggestions, but i'd like to lean away from crypto/security because the point of the MD5 implementation in the plugin is that it isn't used for security purposes. it would raise eyebrows to have an MD5 implementation of something related to security

Comment (shaan1337) [reply src/EventStore.Plugins/Cryptography/ICryptoPlugin.cs]: @timothycoleman hey, yup i agree the name might seem inappropriate. the initial idea for using this name is that we're developing a FIPS plugin, not an MD5 plugin. "crypto" here applies to "FIPS", not to MD5. it just turns out that we have an MD5 implementation in our FIPS cryptography plugin. also, in .NET for historical reasons the `MD5` and `HashAlgorithm` classes are under the `System.Security.Cryptography` namespaces, so it might still be ok to use `Crypto` here.

another reason for also making it a "cryptography" interface and not a "hashing" interface is that i think that there will likely be a need to add additional methods that are not specifically related to hashing to the interface (in the near future):

Using FIPS might come at a performance cost and we do not want to force all our non-FIPS customers to also incur that performance cost as a trade-off for additional security. For example, say FIPS deprecates 1024 bit RSA keys in the near future. We would not necessarily want to change our certificate tooling to generate 2048 bit RSA keys by default. Instead we might want to add some methods to the plugin interface to verify if the server's key/certificate is valid and in the FIPS plugin, we can return an error if it's not compliant and guide the user on how to correctly generate compliant keys/certificates.

To summarize, if we rename the interface to something specific to hashing, it will be difficult to accommodate for changes to the FIPS standard or for FIPS-specific stuff in the same plugin.

Comment (timothycoleman) [reply src/EventStore.Plugins/Cryptography/ICryptoPlugin.cs]: hey @shaan1337 thanks these are good points. i have some thoughts here but no conclusions (its better not to draw conclusions on friday afternoon anyway)

i see also that Crc32 is derived from a different (somewhat amusingly named) class

`public sealed class Crc32 : System.IO.Hashing.NonCryptographicHashAlgorithm`

so i suppose the question will become whether we want this plugin interface to be able to support entirly different hashing mechanisms (in which case it shouldn't depend on HashAlgorithm) or if its really just for replacing one md5 implementation with another

on security related functionality that we might want to add: i think that to accommodate behaviours truly related to security the same dll (what we might call the fips plugin) could implement multiple plugin interfaces (IHashAlgorithmPlugin, ICertificationValidationPlugin ??, etc). they seem like a separate set of behaviours to me, its just that FIPS may have implications in multiple areas

Comment (shaan1337) [reply src/EventStore.Plugins/Cryptography/ICryptoPlugin.cs]: >... (its better not to draw conclusions on friday afternoon anyway)

:smiley: 

> on security related functionality that we might want to add: i think that to accommodate behaviours truly related to security the same dll (what we might call the fips plugin) could implement multiple plugin interfaces (IHashAlgorithmPlugin, ICertificationValidationPlugin ??, etc). they seem like a separate set of behaviours to me, its just that FIPS may have implications in multiple areas

i like this idea. we could then rename the FIPS plugin project to "EventStore.Security.FIPS" or simply "EventStore.FIPS"
the only thing is that when loading them on the server side we would not have a single plugin called "FIPS" but it would be a collection of plugins: "FIPS-MD5", "FIPS-CertificateValidation", etc. and may be slightly confusing for the user but overall it looks fine. if we use a hierarchical plugin structure, then this problem goes away - e.g we could have a type of plugin called `IPluginCollection`

> i see also that Crc32 is derived from a different (somewhat amusingly named) class

thanks, i hadn't noticed this class. from what I understand, the main difference between the abstract class `HashAlgorithm` (used for cryptographic purposes) and the abstract class `NonCryptographicHashAlgorithm` is that the former can be used for block ciphers but the latter cannot (`HashAlgorithm` also implements `ICryptoTransform` which is used for block ciphers)

> so i suppose the question will become whether we want this plugin interface to be able to support entirly different hashing mechanisms (in which case it shouldn't depend on HashAlgorithm) or if its really just for replacing one md5 implementation with another

given the above, it seems to make more sense to make it md5-specific (IMD5Plugin?) and it's also not likely we're going to need to implement hashing algorithms very frequently.

If we keep using `HashAlgorithm` or `ICryptoTransform`, then it keeps things simple on the server-side as right now the block cipher operations are being used on the server: TransformBlock, TransformFinalBlock etc. we could argue that MD5 was once considered a secure cryptographic hash function so it seems fine to continue using `HashAlgorithm` or `ICryptoTransform` for it but one downside is that it might be misused for cryptographic purposes. so, it seems safer to replace it with `NonCryptographicHashAlgorithm` but it would be a bit more work on the server side (to replace the TransformBlock calls, etc. with Append).

Comment (timothycoleman) [reply src/EventStore.Plugins/Cryptography/ICryptoPlugin.cs]: yep agreed, lets call it IMD5Plugin and i'll add a comment below with a simplified rationale

**Comment by timothycoleman** [comment]
Quick summary of decision naming the interface

Context:
- We aren't looking to be able to plugin all sorts of different hash algorithms, that would be too much scope and doesn't seem to be of that much use anyway. In the future if we wanted to support, say, CRC, we could just add it with a configuration option.
- Eventstore uses MD5 for several purposes (all non-crpytographic): chunk checksums, ptable checksums, indexmap, databaselock. Its not always for checksums.

So we avoid calling it `IChecksumPlugin` because it isn't always for checksums and it gives the appearance of having more flexibility than it really does.

We keeps it simple and specific and call it `IMD5Plugin` because it will simply replace the MD5 implementation used within eventstore.

**Comment by timothycoleman** [comment]
looks good, can squash and i'll approve

**Comment by timothycoleman** [comment]
(decided to sqaush it myself since if we get it in it'll help review the other PRs)

**Review by timothycoleman — APPROVED** [review]
:+1:

**Review by thefringeninja — APPROVED** [review]
